crates/stark-felt/src/lib.rs

```diff
@@ -224,8 +303,136 @@ impl TryFrom<&Felt> for NonZeroFelt {
     }
 }
 
+impl FromPrimitive for Felt {
+    fn from_i64(value: i64) -> Option<Self> {
+        Self::from_i128(value as i128)
+    }
+
+    fn from_u64(value: u64) -> Option<Self> {
+        Self::from_u128(value as u128)
+    }
+
+    fn from_i128(value: i128) -> Option<Self> {
+        let mut res = Self(FieldElement::from(&UnsignedInteger::from(
+            value.unsigned_abs(),
+        )));
+        if value.is_negative() {
+            res = -res;
+        }
+        Some(res)
+    }
+
+    fn from_u128(value: u128) -> Option<Self> {
+        Some(Self(FieldElement::from(&UnsignedInteger::from(value))))
+    }
+}
+
+// TODO: we need to decide whether we want conversions to signed primitives
+// will support converting the high end of the field to negative.
+impl ToPrimitive for Felt {
+    fn to_u64(&self) -> Option<u64> {
+        self.to_u128().and_then(|x| u64::try_from(x).ok())
+    }
+
+    fn to_i64(&self) -> Option<i64> {
+        self.to_u128().and_then(|x| i64::try_from(x).ok())
+    }
+
+    fn to_u128(&self) -> Option<u128> {
+        match self.0.representative().limbs {
+            [0, 0, hi, lo] => Some((lo as u128) | ((hi as u128) << 64)),
+            _ => None,
+        }
+    }
+
+    fn to_i128(&self) -> Option<i128> {
+        self.to_u128().and_then(|x| i128::try_from(x).ok())
+    }
+}
+
+macro_rules! impl_from_u128 {
+    ($type:ty) => {
+        impl From<$type> for Felt {
+            fn from(value: $type) -> Self {
+                Self::from_u128(value as u128).expect("conversion from primitive is infallible")
```

Comment:
I'd rather avoid `unsafe` where I can. Even if those alarms are not always justified, it does raise some when it appears. Besides, the `expect` will often be gone after dead code removal, so I wouldn't worry unless I had evidence of the opposite. We could definitely check, but if it can and does remove `assert`s and bound checks it should probably be able to do the same with `expect`. 
That said, implementing `FromPrimitive` in terms of `From` does sound more sensible than the other way around. I only did this this way because I decided on implementing `From` _after_ replacing it with `FromPrimitive` and didn't realize I could do it this way at that point. So, I'll change that in a later commit.